auth/password_hash_test.go

```diff
@@ -64,3 +65,64 @@ func TestSetBcryptCost(t *testing.T) {
 	assert.Equal(t, bcryptDefaultCost, bcryptCost)
 	assert.True(t, bcryptCostChanged)
 }
+
+func BenchmarkCompareHashAndPassword(b *testing.B) {
+	// Get the bcrypt hash of the given password.
+	bcryptHash := func(password []byte) []byte {
+		hash, err := bcrypt.GenerateFromPassword(password, bcryptDefaultCost)
+		require.NoError(b, err)
+		return hash
+	}
+	// hashAndPassword represents the password hash pair.
+	type hashAndPassword struct {
+		hash     []byte
+		password []byte
+	}
+	// Setup cache of size N
+	const numPasswords = 100
+	cachedHashes = NewCache(numPasswords)
+
+	// Generate the test data and warm up the cache if required.
+	generateTestData := func(warmCache bool, key string) []hashAndPassword {
+		hashAndPasswords := make([]hashAndPassword, numPasswords)
+		for i := 0; i < numPasswords; i++ {
+			password := []byte(fmt.Sprintf("%s%d", key, i))
+			hashAndPassword := hashAndPassword{
+				hash:     bcryptHash(password),
+				password: password,
+			}
+			hashAndPasswords[i%numPasswords] = hashAndPassword
+			if warmCache {
+				_ = compareHashAndPassword(hashAndPassword.hash, hashAndPassword.password)
+			}
+		}
+		return hashAndPasswords
+	}
+
+	b.Run("100% cache hit", func(b *testing.B) {
```

Comment:
This subtest is running unexpectedly slow, by many orders of magnitude - `34453444 ns/op` ?
I'd expect something like this test case to be on the order of 100s ns/op.

As a basic sanity check:

```
$ go1.13.4 test -run=- -bench='^BenchmarkAuthKeyCacheContains$' -v -benchmem -count=1 ./auth
goos: darwin
goarch: amd64
pkg: github.com/couchbase/sync_gateway/auth
BenchmarkAuthKeyCacheContains-12    	 4467738	       269 ns/op	     179 B/op	       4 allocs/op
PASS
ok  	github.com/couchbase/sync_gateway/auth	1.688s
```

```golang
func BenchmarkAuthKeyCacheContains(b *testing.B) {
	cache := NewCache(1)
	cache.Put("foo")
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = authKey([]byte("foo"), []byte("bar"))
		_ = cache.Contains("foo")
	}
}
```

I'm fairly sure this is a test issue, rather than implementation, but I can't tell where exactly.

Reply:
There was problem in the test because the cache was shared between multiple benchmarks. Refactored the test and changes are included in latest commit.